File: python/basics/matrix/48_rotate_image.py
#T: O(n**2)
#S: O(1)
#transpose + reflect along vertical
def rotate_image(nums: list[list[int]])->list[list[int]]:
    rows = len(nums)
    cols = len(nums[0])
    #transpose
    for i in range(rows):
        for j in range(i+1, cols):
            nums[j][i],nums[i][j]  = nums[i][j],nums[j][i]
            print("nums: " + str(nums))        

    #reflection
    for i in range(rows):
        for j in range(cols//2):
            nums[i][j], nums[i][cols-j-1]  = nums[i][cols-j-1], nums[i][j]
    return nums


input = [[1,2,3],[4,5,6],[7,8,9]]
print("input:" + str(input))
result = rotate_image(input)
print("result:" + str(result))


input = [[5,1,9,11],[2,4,8,10],[13,3,6,7],[15,14,12,16]]
print("input:" + str(input))
result = rotate_image(input)
print("result:" + str(result))

File: problem_solving/matrix/cpp/48_rotate_image.cpp
/*
matrix
*/
#include <vector>
#include<iostream>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"



void transpose_matrix(std::vector<std::vector<int>>& matrix) {
    int rows = matrix.size();
    int cols = matrix[0].size();
    for (int i = 0; i < rows; ++i) {
        for(int j = i+1; j < cols; ++j )
        {   
            int tmp = matrix[i][j];
            matrix[i][j] = matrix[j][i];
            matrix[j][i] = tmp;
        }
    }
}

void vertical_reflection(std::vector<std::vector<int>>& matrix) {
    int rows = matrix.size();
    int cols = matrix[0].size();
    for(int i = 0; i < rows; ++i) {
        for(int j = 0; j < cols/2; ++j) {
            int tmp = matrix[i][j];
            matrix[i][j] = matrix[i][cols-j-1];
            matrix[i][cols-j-1]=tmp;
        }
    }
}

int main() {
    std::vector<std::vector<int>> matrix = {
        {1,2,3},
        {4,5,6},
        {7,8,9}
    };
    std::cout << "before transpose " << std::endl;
    print_vector_of_vector(matrix);
    transpose_matrix(matrix);
    std::cout << "after transpose " << std::endl;
    print_vector_of_vector(matrix);
    vertical_reflection(matrix);
    std::cout << "after vertical reflection " << std::endl;
    print_vector_of_vector(matrix);
    

    matrix = {
        {5,1,9,11},
        {2,4,8,10},
        {13,3,6,7},
        {15,14,12,16}
    };
    transpose_matrix(matrix);
    vertical_reflection(matrix);
    std::cout << "after vertical reflection " << std::endl;
    print_vector_of_vector(matrix);
    //Output: [[15,13,2,5],[14,3,4,1],[12,6,8,9],[16,7,10,11]]
}
